File: DP/Maximum Sum Increasing Subsequence/MaximumSumIncreasingSubsequence.py
#!/bin/python3
import sys
from copy import copy

fstd = "test2"
sys.stdin = open('%s.txt' % fstd, 'r')
input = sys.stdin.readline

# DP based Python
# implementation of Maximum Sum
# Increasing Subsequence (MSIS)
# problem


# constructMaxSumIS() returns the tupple:
# ( max sum of increasing subsequence, found subsequence )
def constructMaxSumIS(arr):
    N = len(arr)

    # L[i] stores the Maximum Sum Increasing Subsequence that ends with arr[i]
    L = [[] for _ in range(N)]
    # L[0] is equal to arr[0]
    L[0].append(arr[0])

    # msis[i] stores the Maximum Sum of Increasing Subsequence that
    # ends with arr[i]
    msis = [0] * N
    # Initialize msis values
    msis[0] = arr[0]

    # Compute maximum sum
    # values in bottom up manner
    for i in range(1, N):
        for j in range(i):
            if (arr[i] > arr[j]) and (msis[i] < msis[j]):
                msis[i] = msis[j]
                L[i] = copy(L[j])
        L[i].append(arr[i])
        msis[i] += arr[i]

    # Find max
    maxIndex = max(range(N), key=lambda i: msis[i])
    return (msis[maxIndex], L[maxIndex])


if __name__ == "__main__":
    print("Input sequence:")
    A = []
    while (True):
        line = input().strip()
        try:
            num = int(line)
        except ValueError:
            break
        A.append(num)
    print(A)
    res = constructMaxSumIS(A)
    print("Max sum is {}".format(res[0]))
    print("Found sequence is: {} ".format(res[1]))

File: DP/Maximum Sum Increasing Subsequence/MaximumSumIncreasingSubsequence.cpp
// sol.cpp : Defines the entry point for the console application.
//

//#include "stdafx.h"
#include <bits/stdc++.h>
/* Dynamic Programming solution to construct Maximum Sum Increasing Subsequence */

#include <cmath>
#include <cstdio>
#include <fstream>
#include <string>
#include <iostream>
#include <vector>
#include <utility>
#include <algorithm>
using namespace std;


typedef vector<int> VI;

// Function to construct Maximum Sum Increasing Subsequence
VI calcMaxSumIS(VI arr)
{
	int n = arr.size();

	// L[i] stores the Maximum Sum Increasing Subsequence that ends with arr[i]
	vector<VI> L(n);

	// msis[i] stores the Maximum Sum of Increasing Subsequence that
	// ends with arr[i]
	VI msis(n);

	// L[0] is equal to arr[0]
	L[0].push_back(arr[0]);
	msis[0] = arr[0];

	// start from index 1
	for (int i = 1; i < n; i++)
	{
		// for every j less than i
		for (int j = 0; j < i; j++)
		{
			// L[i] = {MaxSum(L[j])} + arr[i] where j < i and arr[j] < arr[i]
			if ((arr[i] > arr[j]) && (msis[i] < msis[j]))
			{
				L[i] = L[j];
				msis[i] = msis[j];
			}
		}
		// L[i] ends with arr[i]
		L[i].push_back(arr[i]);
		msis[i] += arr[i];

		// L[i] now stores Maximum Sum Increasing Subsequence of arr[0..i]
		// that ends with arr[i]
	}

	int maxIndex = 0;
	int max = msis[0];

	// find max
	for (int i = 1; i < n; i++)
		if (msis[i] > max)
		{
			max = msis[i];
			maxIndex = i;
		}

	// max will contain result
	return L[maxIndex];
}

int main() {
	ifstream infile("test1.txt");
	if (infile)
	{
		VI arr;
		string s;
		int a;
		while(true) {
			try {
				getline(infile,s);
				a = stoi(s);
			}
			catch (...) {
				break;
			}
			arr.push_back(a);
		}
		auto res= calcMaxSumIS(arr);
		for (auto r : res)
			cout << r << " ";
		cout << endl;
	}
}